### PR TITLE
TriangulationDescription::Utilities::create_description_from_triangulation(): allow subcommunicator in p::d::T

### PR DESCRIPTION
**Pull request by peterrum**

### CODE REVIEW

File: tests/fullydistributed_grids/repartitioning_05.cc [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+  for (const auto &cell : tria_in.active_cell_iterators())
+    if (cell->is_locally_owned())
+      partition[cell->global_active_cell_index()] =
+        cell->global_active_cell_index() * n_partitions /
+        tria_in.n_global_active_cells();
+
+  partition.update_ghost_values();
+
+  return partition;
+}
```
Comment (peterrum) [inline tests/fullydistributed_grids/repartitioning_05.cc:98]: @kronbichler This is how a partitioning policy might look like in this context. "Simply" replace the communicator in the partitioner...

File: tests/fullydistributed_grids/repartitioning_05.cc [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+  int color = rank < size;
+
+  MPI_Comm sub_comm;
+  MPI_Comm_split(comm, color, rank, &sub_comm);
+
+  if (rank < size)
+    return sub_comm;
+  else
+    return MPI_COMM_NULL;
```
Comment (kronbichler) [inline tests/fullydistributed_grids/repartitioning_05.cc:54]: Before I go on with the review, let me ask a question (I also wondered when implementing related functionality): Aren't you throwing away the ranks `>= size`, but without freeing the associated communicator? I don't know all MPI semantics, but I would have expected that I need to call `MPI_Comm_free(&sub_comm)` in the `else` branch. If MPI wants to free all communicators coming out of a `MPI_Comm_split` at the same time, then we would need to return the `sub_comm` part and pick up a different decision.

Comment (peterrum) [reply tests/fullydistributed_grids/repartitioning_05.cc]: Fixed.

File: source/grid/tria_description.cc [hunk 3]
```diff
@@ -745,6 +745,11 @@ namespace TriangulationDescription
       const Triangulation<dim, spacedim> &              tria,
       const LinearAlgebra::distributed::Vector<double> &partition)
     {
+#ifdef DEAL_II_WITH_MPI
+      if (tria.get_communicator() == MPI_COMM_NULL)
+        AssertDimension(partition.local_size(), 0);
+#endif
+
```
Comment (kronbichler) [inline source/grid/tria_description.cc:752]: Do you want to add a short discussion about the possibilities for communicators to the header file?

**Review by kronbichler — APPROVED** [review]

File: include/deal.II/grid/tria_description.h [hunk 4]
```diff
@@ -455,7 +455,12 @@ namespace TriangulationDescription
      * CellAccessor::global_active_cell_index()). This function allows to
      * repartition distributed Triangulation objects.
      *
-     * @note The communicator is extracted from the Triangulation @p tria.
+     * @note The communicator is extracted from the vector @p partition.
+     *
+     * @note The triangulation @p tria can be set up on a subcommunicator of the
+     *   communicator of @p partitioner. All processes that are not part of that
+     *   subcommunicator need to set up the local triangulation with the
+     *   special-purpose communicator MPI_COMM_NULL.
```
Comment (peterrum) [inline include/deal.II/grid/tria_description.h:463]: Done!

**Review by kronbichler — APPROVED** [review]

**Pull request merged**